lib/linguist/languages.yml

```diff
@@ -3224,11 +3224,16 @@ MoonScript:
   language_id: 238
 Motorola 68K Assembly:
   type: programming
-  group: Assembly
+  aliases:
+  - m68k
   extensions:
-  - ".X68"
+  - ".asm"
+  - ".i"
+  - ".inc"
+  - ".s"
```

Comment:
Are there any registers or opcodes unique to Motorola we can use to disambiguate assembly files with?

We're *definitely* going to need some heuristics for `.asm` and `.inc`. The latter of which is particularly important because it sees very general use across a range of unrelated) languages…

Reply:
M68k assembly is easily distinguished from other assembly languages, both by registers and opcodes. 

How would such a heuristic look?

Reply:
A regular expression; I'm happy to write it for you, provided you give me the names of substrings guaranteed (or highly unlikely) to appear in the source code of any other assembler language.

Reply:
`(?im:moveq\b.*?d\d|move\.[bwl]\s+.*\b[ad]\d|movem\.[bwl]\b|btst\b|dbra\b)`

Reply:
Would it be reasonable to limit the `moveq` heuristic to match two registers (one address, one data)? From what I hear, 68k is unique for differentiating between the two.

If so, we could try this:

~~~regexp
(?xi)
	# Mnemonic
	\b moveq (\.l)? \s+
	
	# Address
	\#( \$ -? [0-9a-f]{1,3}
	  |    %  [0-1]{1,8}
	  |    -? [0-9]{1,3}
	  )
	, \s*
	
	# Register
	d[0-7] \b
~~~

When writing heuristics, it's best to be as specific as possible; anything which doesn't match is passed down to the (less accurate) classification techniques.

Reply:
If I read that regex correctly, it looks perfect. I'm just not as good at writing them.

Reply:
Credit for this expression belongs to @zerkman, since it's taken from the [`language-m68k`](https://github.com/zerkman/language-m68k/blob/f7f7bf11c322976bd9fd00a0d94a1cfc3d1c6c8f/grammars/m68k.cson#L144) grammar we're using to highlight 68k on GitHub (I did clean it up and remove some redundant syntax for clarity).

I've amended the other parts of that expression to use other bits of that grammar, bringing us down to:

~~~regexp
(?xim:
	# Mnemonic
	\b moveq (\.l)? \s+
	
	# Address
	\#( \$ -? [0-9a-f]{1,3}
	  |    %  [0-1]{1,8}
	  |    -? [0-9]{1,3}
	  )
	, \s*
	
	# Register
	d[0-7] \b
	
	| ^ \s* move     (\.[bwl])? \s+ (sr|usp), \s* [^\s]+
	| ^ \s* movem     \.[bwl]  \b
	| ^ \s* move[mp] (\.[wl])? \b
	| ^ \s* btst  \b
	| ^ \s* dbra  \b
)
~~~

Notice that I've anchored the remaining parts to match at the beginning of a line (with or without indentation). This reduces the risk of incorrectly matching part of a comment in an unrelated file. For the same reason, you'll notice I avoid using wildcards when possible (`.*`).

Reply:
@idrougge If the above revisions look good to you, then the changes to make to `heuristics.yml` are below.

I'll still need to test them thoroughly on my end, as well as investigate any possible formats using the `.i` extension that we've not registered yet.

<details><summary><b>Click to show diff</b></summary>

~~~diff
--- heuristics.yml	2019-10-03 21:45:48.000000000 +1000
+++ heuristics.yml	2019-10-03 22:08:15.000000000 +1000
@@ -49,8 +49,12 @@
   rules:
   - language: ActionScript
     pattern: '^\s*(package\s+[a-z0-9_\.]+|import\s+[a-zA-Z0-9_\.]+;|class\s+[A-Za-z0-9_]+\s+extends\s+[A-Za-z0-9_]+)'
   - language: AngelScript
+- extensions: ['.asm']
+  rules:
+  - language: Motorola 68K Assembly
+    named_pattern: m68k
 - extensions: ['.asc']
   rules:
   - language: Public Key
     pattern: '^(----[- ]BEGIN|ssh-(rsa|dss)) '
@@ -191,8 +195,10 @@
     pattern: '\A\s*[{\[]'
   - language: Slice
 - extensions: ['.inc']
   rules:
+  - language: Motorola 68K Assembly
+    named_pattern: m68k
   - language: PHP
     pattern: '^<\?(?:php)?'
   - language: SourcePawn
     pattern: '^public\s+(?:SharedPlugin(?:\s+|:)__pl_\w+\s*=(?:\s*{)?|(?:void\s+)?__pl_\w+_SetNTVOptional\(\)(?:\s*{)?)'
@@ -383,8 +389,12 @@
   - language: Rust
     pattern: '^(use |fn |mod |pub |macro_rules|impl|#!?\[)'
   - language: RenderScript
     pattern: '#include|#pragma\s+(rs|version)|__attribute__'
+- extensions: ['.s']
+  rules:
+  - language: Motorola 68K Assembly
+    named_pattern: m68k
 - extensions: ['.sc']
   rules:
   - language: SuperCollider
     pattern: '(?i:\^(this|super)\.|^\s*~\w+\s*=\.)'
@@ -486,7 +496,14 @@
   - '^[ \t]*(private|public|protected):$'
   - 'std::\w+'
   fortran: '^(?i:[c*][^abd-z]|      (subroutine|program|end|data)\s|\s*!)'
   key_equals_value: '^[^#!;][^=]*='
+  m68k:
+  - '(?im)\bmoveq(?:\.l)?\s+#(?:\$-?[0-9a-f]{1,3}|%[0-1]{1,8}|-?[0-9]{1,3}),\s*d[0-7]\b'
+  - '(?im)^\s*move(?:\.[bwl])?\s+(?:sr|usp),\s*[^\s]+'
+  - '(?im)^\s*movem\.[bwl]\b'
+  - '(?im)^\s*move[mp](?:\.[wl])?\b'
+  - '(?im)^\s*btst\b'
+  - '(?im)^\s*dbra\b'
   objectivec: '^\s*(@(interface|class|protocol|property|end|synchronised|selector|implementation)\b|#import\s+.+\.h[">])'
   perl5: '\buse\s+(?:strict\b|v?5\.)'
   perl6: '^\s*(?:use\s+v6\b|\bmodule\b|\b(?:my\s+)?class\b)'
~~~

</details>

Reply:
I tested the `^ \s* move     (\.[bwl])? \s+ (sr|usp), \s* [^\s]+` line, and it only catches moves to `sr` or `usp`, not moves to any given register. I feel that the Motorola syntax of `move.size with source or destination as a register named Dn or An is sufficiently dissimilar to other assembly syntaxes to avoid confusion with other assembly languages while also catching even the shortest snippet.
Testing may prove the totality of heuristics to still be sufficient to catch all m68k assembly sources. 

`.i `, like `.inc`, `.asm` or `.s` is used by assemblers on most platforms AFAIK.

Reply:
In that case, the third line of the `m68k` heuristic becomes:

~~~yaml
- '(?im)^\s*move\.[bwl]\s+.*\b[ad]\d'
~~~

I'll update the diff I just posted.

> `.i` , like `.inc`, `.asm` or `.s` is used by assemblers on most platforms AFAIK.

There are currently [11,273,157 `.i` files](https://github.com/search?q=extension%3Ai+NOT+nothack&type=Code) publicly indexed on GitHub. Surely there must be other formats hidden out there...

Reply:
That regex looks fine for heuristics.

A quick glance at those results indicate that a lot of `.i` files are SWIG files, which may need a language definition of their own.